docs/developer/architecture/ci.rst

```diff
@@ -0,0 +1,122 @@
+CI
+==
+
+Pre-merge
+---------
+
+What should be tested in pre-merge on every ``user/*``, ``feature/*``,
+``improvement/*``, ``bugfix/*``, ``w/*`` branches. The pre-merge test should
+not long too much time (less than 40 minutes is great) so we can't
+test everything in pre-merge, we should only test building of the product
+and check that product still usable.
+
+- Building tests
+
+  - Build
+  - Lint
+  - Unit tests
+
+- Installation tests
+
+  - Simple install RHEL
+  - Simple install CentOs + expansion
+
+When merging several pull requests at the same time, given that we are on a
+queue branch (``q/*``), we may want to have or we may require additional tests
+as combinaison of several PRs could have larger impact than all individual PR.
+
+- Simple upgrade/downgrade
+
+.. _post-merge:
+
+Post-merge
+----------
+
+On each and every ``development/2.*`` branches we want to run complex tests
+to ensure that the current product continues to work.
+
+Nightly
+^^^^^^^
+
+- Upgrade, downgrade tests:
+
+  - For development branch just before
+
+    e.g.: on ``development/2.x`` test upgrade from ``development/2.(x-1)``
+    and downgrade to ``development/2.(x-1)``
+
+    - Build branch ``development/2.(x-1)``
+    - Tests:
+
+      - Single node test
+      - Complex deployment test
```

Comment:
Yes just need to define complex deployment

Reply:
I was wondering if you meant "simple + complex for **each** [up/down]grade scenario". If yes, I think it's overkill. Either just run it in single node or multiple nodes, but I don't see much benefits in running all those cases nightly. Maybe have full coverage on weekly builds.

Reply:
Single node upgrade/downgrade is REALLY different from a full cluster with several etcd upgrade/downgrade

Reply:
Agreed, both cases need to be tested. But should they be tested for each scenario? We're talking two architectures, for six scenarii (3 upgrades, 3 downgrades), for seven development branches. This means we're running 84 install + [up/down]grade every night.

Reply:
IMO yes but it's not 84 as you do not test it for dev/2.0 and also for the moment we did not have released version for 2.0 2.1 2.2 2.3 2.5 2.6 but anyway IMO we should test everything nightly, 
For me weekly we should do long test like for example creating a mixed OS cluster and doing several test for the same env like:
- install bootstrap
- Upgrade
- Downgrade
- expansion
- Upgrade
- Crash bootstrap, restore a new one
- expansion
- Upgrade
- Downgrade
So we test "complex stuff" like that backup/restore still working after upgrade/downgrade, expansion still working after upgrade/downgrade and after restore, ...

Anyway if we have too much install to do nightly we can think about skipping commit already tested (mean if a branch did not change since the day before we do not test it)

Reply:
OK, let's do everything until we find problems.